i18n: use IcuMessage objects instead of string IDs

brendankenny opened the pull request:
As discussed in #10614 and offline, this moves us to using objects for localizable strings. This has several advantages (listed in that issue) and it turns out few disadvantages.

Proposed object:
```ts
type IcuMessage = {
  /** The id locating this message in the locale message json files. */
  i18nId: string;
  /** The dynamic values, if any, to insert into the localized string. */
  values?: Record<string, string | number>;
  /** A formatted version of the string, usually as found in a file's `UIStrings` entry, used as backup. */
  formattedDefault: string,
};
```
Biggest change from the proposal is the `formattedDefault` which will always be there. This is basically what `swap-locale` already does (uses the string already in the LHR as backup if lighthouse has changed enough that the message ID no longer refers to a translated string anymore) but uses it for allowing serialization at any stage, as well as for the current UIStrings backup during development (allowing use of a string without having to run `yarn update:sample-json` after every edit).

The PR changes are a little spread out, but the substantial changes are in
- `i18n.js`
- `i18n.d.ts`
- 'swap-locale.js`
- `runner.js`
- `details-renderer.js`
- `audit.d.ts`, `audit-details.d.ts`, `lhr.d.ts`
more or less in that order for significance of the changes.

brendankenny commented on lighthouse-cli/run.js:
```diff
@@ -20,7 +20,7 @@ const assetSaver = require('../lighthouse-core/lib/asset-saver.js');
 const open = require('open');
 
-/** @typedef {import('../lighthouse-core/lib/lh-error.js')} LighthouseError */
+/** @typedef {Error & {code: string, friendlyMessage?: string}} ExitError */
```
wasn't worth pretending this was a `LighthouseError` since either it's a regular `Error` or we construct it as just an object down on line 224 :)

brendankenny commented on lighthouse-core/config/config.js:
```diff
@@ -664,22 +664,6 @@ class Config {
-   * @param {LH.Config.Json} config
-   * @return {Array<{id: string, title: string}>}
-   */
-  static getCategories(config) {
```
this is only used in two tests now (maybe after the extension was removed?), so it seemed easier just to delete than to deal with localization

brendankenny commented on lighthouse-core/lib/i18n/swap-locale.js:
```diff
@@ -65,21 +66,37 @@ function swapLocale(lhr, requestedLocale) {
+      // If the path isn't valid or the value isn't a string, there's no point in trying to replace it.
+      const originalString = _get(lhr, path);
+      if (typeof originalString !== 'string') {
+        // TODO(bckenny): what to do in this case? Ignore?
```
@paulirish have we ever cared about `missingIcuMessageIds`? They're kind of helpful for tests, but I'm not sure if there's much value in expanding the kinds of errors we log to an array no one looks at :)

In this case the two TODOs are what to do with
- `icuMessagePath` that doesn't actually refer to a string in the LHR. This is either a bug in path generation (so should be caught by tests) or a malformed LHR (in which case...what could we even do), so ignoring may be fine?
- We log missing IDs, but what if the ID exists but the number or type of replacement values has changed? I think it would just throw on master, but do we want to put that case into `missingIcuMessageIds` as well?

paulirish replied:
> @paulirish have we ever cared about missingIcuMessageIds? 

i don't. it came up while writing this but it's never been a problem in reality.

brendankenny commented on lighthouse-core/report/html/renderer/details-renderer.js:
```diff
@@ -20,6 +20,14 @@
+/** @typedef {LH.FormattedIcu<LH.Audit.Details.Opportunity>} OpportunityTable */
+/** @typedef {LH.FormattedIcu<LH.Audit.Details.OpportunityColumnHeading>} OpportunityTableHeading */
+/** @typedef {LH.FormattedIcu<LH.Audit.Details.Table>} Table */
+/** @typedef {LH.FormattedIcu<LH.Audit.Details.TableColumnHeading>} TableHeading */
```
This is my least favorite part of this change and I haven't come up with a nice solution yet. Unlike with Audit product/result, we don't have a good split on audit details (the same types are used in audits and in the LHR).

The types are all correct on the LHR, but if you want to refer to the type in isolation (and don't want to do `LH.Result['audits'][number]['details']['whatever']['etc']['etc']`) you need a reference to the type directly like this.

Other options:
- we have some sort of audit details product/result split?
- move these under `LH.Result` namespace?

patrickhulce replied:
> move these under LH.Result namespace?

That seems reasonable but this works too.

brendankenny commented on lighthouse-core/runner.js:
```diff
@@ -149,15 +143,20 @@ class Runner {
-      // i18n error strings
-      err.friendlyMessage = i18n.getFormatted(err.friendlyMessage, settings.locale);
+      // i18n LighthouseError strings.
+      if (err.friendlyMessage) {
```
turns out these aren't always LHErrors and we were often formatting `undefined` :)

paulirish commented on lighthouse-core/lib/i18n/i18n.js:
```diff
@@ -268,56 +265,45 @@ function _preformatValues(icuMessage, messageFormatter, values) {
+ * @return {string}
  */
-function _formatIcuMessage(locale, icuMessageId, uiStringMessage, values = {}) {
+function _localizeIcuMessage(icuMessage, locale) {
```
i like the new split between `_localizeIcuMessage` and `_formatMessage`

paulirish commented on lighthouse-core/lib/i18n/i18n.js:
```diff
@@ -358,143 +344,140 @@ function getRendererFormattedStrings(locale) {
  * @param {string} filename
  * @param {Record<string, string>} fileStrings
  */
 function createMessageInstanceIdFn(filename, fileStrings) {
```
can we rename this to `createIcuMessageFn` internally here.. we can still expose it in module.exports as the long name (for now)

brendankenny replied:
> can we rename this to `createIcuMessageFn` internally here.. we can still expose it in module.exports as the long name (for now)

should both names be exposed and we convert all Lighthouse audits over to using `createIcuMessageFn`?

patrickhulce commented on lighthouse-core/audits/byte-efficiency/render-blocking-resources.js:
```diff
@@ -208,7 +208,7 @@ class RenderBlockingResources extends Audit {
     const {results, wastedMs} = await RenderBlockingResources.computeResults(artifacts, context);
 
-    let displayValue = '';
+    let displayValue;
```
these are inferred correctly, right? in tsc playground it seems to work out out OK but some of the assertions we're removing later on in jsdoc makes me think some of them were necessary at one point in time for our checkJs setup...

https://www.typescriptlang.org/play/#code/DYUwLgBAZg9jDcAoRBLKEAUBZAhmAFgHQBOOAdgCYwC2GAlBAHwQAMhArAwN6IR-RwIAXghcAHgC4IARgA0EAJ5SATAF9E65AGMYZAM4xQhYDADmGWDEIAvOkA

patrickhulce commented on lighthouse-core/config/config-helpers.js:
```diff
@@ -34,6 +35,14 @@ const mergeOptionsOfItems = function(items) {
+ * @param {unknown} value
+ * @return {value is string|LH.IcuMessage}
+ */
+function isStringOrIcuMessage(value) {
```
should this just live on i18n or you feel like it won't come up again?

brendankenny replied:
> should this just live on i18n or you feel like it won't come up again?

it seemed like just an internal helper fn for config (since the inline check was too long for the line and repeated 3x), so it's just scoped to the module, but I don't have strong feelings about keeping it locked away.

patrickhulce commented on lighthouse-core/lib/i18n/README.md:
```diff
@@ -213,13 +213,15 @@ CTC is a name that is distinct and identifies this as the Chrome translation for
 
 # Appendix
 
+TODO(bckenny): update
```
is this TODO as part of this PR or later? :)

brendankenny replied:
> is this TODO as part of this PR or later? :)

part of this PR :) I'll try to figure out what I thought should be added

patrickhulce commented on lighthouse-core/lib/i18n/i18n.js:
```diff
@@ -6,19 +6,17 @@
 /** @typedef {import('intl-messageformat-parser').ArgumentElement} ArgumentElement */
 
 const LH_ROOT = path.join(__dirname, '../../../');
-const MESSAGE_INSTANCE_ID_REGEX = /(.* \| .*) # (\d+)$/;
```
this thing is now completely eliminated, correct?

brendankenny replied:
> this thing is now completely eliminated, correct?

yes, now it's just a check for the locale json file keys (it could also be inlined in the single place it's used now)

patrickhulce commented on lighthouse-core/lib/i18n/i18n.js:
```diff
@@ -202,12 +199,12 @@ function collectAllCustomElementsFromICU(icuElements, seenElementsById = new Map
- * @param {string} icuMessage
  * @param {MessageFormat} messageFormatter
  * @param {Readonly<Record<string, string | number>>} values
+ * @param {string} icuMessage Used for clear error logging.
```
should we rename this param now that `icuMessage` means `that object thing returned from a string we want localized`

brendankenny replied:
> should we rename this param now that `icuMessage` means `that object thing returned from a string we want localized`

I'll use `lhlMessage` since it's in that format? Or should it be `rawMessage` or..something?

patrickhulce commented on lighthouse-core/lib/i18n/i18n.js:
```diff
@@ -358,143 +344,140 @@ function getRendererFormattedStrings(locale) {
-
-    _icuMessageInstanceMap.set(icuMessageId, icuMessageInstances);
-
-    return `${icuMessageId} # ${indexOfInstance}`;
```
we can eliminate the `# ${index}` part of this because we keep the values to format in the object itself now right?

patrickhulce commented on lighthouse-core/lib/i18n/i18n.js:
```diff
@@ -358,143 +344,140 @@ function getRendererFormattedStrings(locale) {
+    }
+    for (const value of Object.values(values)) {
+      if (typeof value !== 'string' && typeof value !== 'number') {
+        return false;
```
what are the consequences if we're accidentally passing an undefined or null or something in here? presumably it should be impossible due to types, but what if :)

brendankenny replied:
> what are the consequences if we're accidentally passing an undefined or null or something in here? presumably it should be impossible due to types, but what if :)

it'll end up throwing in `_preformatValues` when it tries to insert the null/undefined into the message but it only allows strings (inserted unchanged) or numbers (inserted formatted).

We could do a check in `createMessageInstanceIdFn`/`createIcuMessageFn` so it'll throw closer to the root cause (immediately on calling `_str()`) but throwing in `_preformatValues` is what happens in the current LH too.

patrickhulce replied:
got it, yeah no need for changes here, just wondering if this introduced anything new to worry about 👍

patrickhulce commented on types/i18n.d.ts:
```diff
@@ -0,0 +1,79 @@
+     * Heavy handed and requires more type checks, so prefer explicitly setting
+     * properties to include `LH.IcuMessage` over this helper if possible.
+     */
+    type Icu<T> = T extends IcuMessage ? T :
```
how do you feel about calling this `UnformattedIcu` or `RawIcu` or something?

it was a bit weird seeing this in usage until I came here to see it sitting in direct contrast to `FormattedIcu`

brendankenny replied:
> how do you feel about calling this `UnformattedIcu` or `RawIcu` or something?
> 
> it was a bit weird seeing this in usage until I came here to see it sitting in direct contrast to `FormattedIcu`

`RawIcu` sounds good to me since `UnformattedIcu` is a little visually confusable with `FormattedIcu`, but happy to bikeshed more now that it's been 5 months since this comment :)

patrickhulce commented on lighthouse-core/test/lib/i18n/i18n-test.js:
```diff
@@ -32,17 +32,20 @@ describe('i18n', () => {
 
   describe('#createMessageInstanceIdFn', () => {
-    it('returns a string reference', () => {
+    it('returns a IcuMessage reference', () => {
```
```suggestion
    it('returns an IcuMessage reference', () => {
```

patrickhulce commented on lighthouse-core/test/lib/i18n/swap-locale-test.js:
```diff
@@ -52,6 +52,7 @@ describe('swap-locale', () => {
         redirects: {
           id: 'redirects',
           title: 'Avoid multiple page redirects',
+          doesntExist: 'A string that does not have localized versions',
```
should we add a different path to something that doesn't exist then? feels like this is a duplicate test of fakeaudit title otherwise

paulirish commented:
TODO: add test for serializability ;)

brendankenny reviewed (commented):
still TODO:
- [ ] `createIcuMessageFn` [question](https://github.com/GoogleChrome/lighthouse/pull/10630/files#r416261011)
- [ ] update `i18n/readme.md`, apparently
- [ ] serialization check

as mentioned in the meeting today, several more people have spent time with localization and audit-details, etc since this PR was opened, so feel free to weigh in on the design!

patrickhulce commented on lighthouse-core/lib/i18n/i18n.js:
```diff
@@ -369,143 +354,140 @@ function getRendererFormattedStrings(locale) {
-}
+  // formattedDefault is required.
+  if (typeof formattedDefault !== 'string') {
+    return false;
```
way to go codecov, it's like a burndown list of test cases when it's a utility module 👍

patrickhulce reviewed (commented):
the bones still look great!

brendankenny commented on types/audit.d.ts:
```diff
@@ -66,23 +66,25 @@ declare global {
       wastedPercent?: number;
     }
 
+    // TODO: consider making some of the `string | IcuMessage` into just `IcuMessage` to require i18n.
```
> is this the main reason we're doing `string | IcuMessage` manually here instead of using the `RawIcu` type?

Yeah, and the ergonomics are also just more straightforward to use, since `RawIcu<>` makes it an indirect step when using in vscode...you can't just see the type on the object, it'll look like `LH.RawIcu<{text: string}>` or `LH.RawIcu<ProductBase>` (and if you click through to `ProductBase` you would only see `string`), so it would only be when accessing the property directly that you would see the type.

(There's [a comment](https://github.com/GoogleChrome/lighthouse/pull/10630/files#diff-ec1b2a8c6b3014e062be782e9bfe51fcR32-R33) on the `RawIcu` type that points to prefer making properties `| LH.IcuMessage` instead of using it for this reason)

I tried making more of these `LH.IcuMessage` only, and I think for every one of them there was some audit that was currently breaking the rules and had a string instead of a localized message, but that was in April, maybe it's better. There's also the question of if these types are just for us, or also for folks writing plugins/custom audits, and they might not want to bother with localization yet, so not allowing `string` would break them.

brendankenny commented on lighthouse-core/test/runner-test.js:
```diff
@@ -155,6 +156,63 @@ describe('Runner', () => {
       });
     });
+
+    it('serializes IcuMessages in gatherMode and is able to use them in auditMode', async () => {
```
kind of a monster sized test because it does a gatherMode run _and_ an auditMode run, but I also didn't really want to split because it's annoying when tests have interdependencies, so 🤷

patrickhulce replied:
It might make sense to split the three -G, -A, and `loadArtifacts` components separately since we're asserting the exact shape of the artifacts already? 

I would prioritize the e2e flow though like you did here anyway 🤷

brendankenny commented on lighthouse-core/test/runner-test.js:
```diff
@@ -155,6 +156,63 @@ describe('Runner', () => {
+      class WarningAndErrorGatherer extends Gatherer {
+        afterPass(passContext) {
+          const warning = str_(i18n.UIStrings.displayValueByteSavings, {wastedBytes: 2222});
+          passContext.LighthouseRunWarnings.push(warning);
```
lherrors have actually been serializing fine since #9397, it's other strings like runWarnings that have been causing the -G/-A problems (see https://github.com/GoogleChrome/lighthouse/issues/9269#issuecomment-512622306), so this test checks both are working.

brendankenny commented on lighthouse-core/test/runner-test.js:
```diff
@@ -155,6 +156,63 @@ describe('Runner', () => {
+      expect(lhr.audits['dummy-audit']).toMatchObject({
+        scoreDisplayMode: 'error',
+        // eslint-disable-next-line max-len
+        errorMessage: 'Required WarningAndErrorGatherer gatherer encountered an error: UNSUPPORTED_OLD_CHROME',
```
should probably follow up this PR and make the automatic audit errors that come from artifact errors use `friendlyMessage` instead of just `message` (which is [set to just the `code`](https://github.com/GoogleChrome/lighthouse/blob/c6e753c191ac96300a931f2a4c434083590b265f/lighthouse-core/lib/lh-error.js#L111)), because a lot of context is lost here. Downside: error messages get a lot longer.

brendankenny commented:
outstanding TODOs are complete, ready for review :)

brendankenny commented:
@connorjclark smoke-devtools times out no matter how many times I run this (and it's rebased against latest). Any ideas?

patrickhulce commented on lighthouse-core/test/runner-test.js:
```diff
@@ -155,6 +156,63 @@ describe('Runner', () => {
+        name: 'LHError',
+        code: 'UNSUPPORTED_OLD_CHROME',
+        // eslint-disable-next-line max-len
+        friendlyMessage: expect.toBeDisplayString(`This version of Chrome is too old to support 'VRML'. Use a newer version to see full results.`),
```
😆

patrickhulce commented:
Are there devtools changes required to make this work with the timeouts? Nothing immediately comes to mind, but definitely suspicious...

I love the direction this takes us! Much more self-contained and explicit about what's i18n'd and what's not 💯 👍

connorjclark commented:
The recent changes in devtools broke Lighthouse in release and locally. I think they've reverted it since.

I don't think there's a way to clear the github action cache... you might need to add a nonce comment in the devtools.yml, to trigger a new cache.

brendankenny commented:
hmm, it's the same key and commit that's passed on the latest commit on master, but caches are hard, so I'll try it :)

connorjclark commented:
I ran it locally and got this error:

![image](https://user-images.githubusercontent.com/4071474/93507703-ae138180-f8e3-11ea-936b-b30af1c4783d.png)

brendankenny commented:
weird, but that's great, that really is something that needs to be updated. Any ideas on how we could pipe startup errors like that back to github actions instead of timing out?

brendankenny commented:
OK:
- moved `isStringOrIcuMessage` to `i18n.js` so `config-helpers` and `config-plugin` can use it
- `config-plugin` now allows `IcuMessage` for the relevant config strings
- added a `config-plugin` test with a localizable plugin
- temporarily switched all the `string | IcuMessage` types to just `IcuMessage` and verified there aren't any other places that should account for `IcuMessage` but currently don't

connorjclark commented on lighthouse-core/test/chromium-web-tests/run-web-tests.sh:
```diff
@@ -84,7 +84,7 @@ cp "$DEVTOOLS_PATH/test/webtests/http/tests/devtools/lighthouse/"*-expected.txt
   # Print failure diffs to stdout.
   find "$LH_ROOT/.tmp/layout-test-results/retry_3" -name '*-diff.txt' -exec cat {} \;
-  echo "❌❌❌ webtests failed. to rebaseline run: yarn update:webtests ❌❌❌"
+  echo "❌❌❌ webtests failed. to rebaseline run: yarn update:test-devtools ❌❌❌"
```
bingo

connorjclark commented on third-party/chromium-webtests/webtests/http/tests/devtools/lighthouse/lighthouse-emulate-run-expected.txt:
```diff
@@ -25,5 +25,5 @@ ViewportDimensions: {
 
 
-content-width: pass 
+content-width: pass undefined
```
hmm?
![image](https://user-images.githubusercontent.com/4071474/93530076-94cdfd80-f902-11ea-87ae-66e9e2203501.png)

brendankenny replied:
`audit.explanation` is optional, so can be `undefined` :)

It was the empty string before...not sure that's better :)

patrickhulce replied:
seems like we should only put `audit.explanation` into the result when the audit is failing 🤔

connorjclark replied:
can you add to the list? I also want to pull this into a fn in LighthouseTestRunner (also used in other tests)

patrickhulce commented:
new changes LGTM too, so I'll let my review stand. Yay devtools tests for catching things at PR time 😃  🎉
